### PR TITLE
IPC4: add llp reading support

### PR DESCRIPTION
**Pull request by RanderWang**
Linear link position (llp) is updated in llp register by GP-DMA hardware. Now it is enabled by SOF FW but not used.
This patch set up a path to read it and update it to host driver.

### CODE REVIEW

File: src/include/sof/lib/dma.h [hunk 1]
```diff
@@ -149,6 +149,8 @@ struct dma_chan_status {
 	uint32_t w_pos;
 	uint32_t r_pos;
 	uint32_t timestamp;
+
+	void *private;
```
Comment (lgirdwood) [inline src/include/sof/lib/dma.h:153]: Can you add a comment here describing this is for ipc dependent position info. I would rename to ```ipc_posn_data```.

File: src/audio/copier.c [hunk 2]
```diff
@@ -719,6 +719,84 @@ static int copier_set_large_config(struct comp_dev *dev, uint32_t param_id,
 	}
 }
 
+static inline void divide_u64_to_u32s(uint64_t val, uint32_t *val_l, uint32_t *val_h)
```
Comment (lgirdwood) [inline src/audio/copier.c:722]: lets rename this since we are not doing a mathematical division.

File: src/audio/copier.c [hunk 3]
```diff
@@ -719,6 +719,84 @@ static int copier_set_large_config(struct comp_dev *dev, uint32_t param_id,
...
+{
+	struct copier_data *cd = comp_get_drvdata(dev);
+	struct sof_ipc_stream_posn posn;
+	struct ipc4_llp_reading_extended llp_ext;
+	struct ipc4_llp_reading llp;
+
+	switch (param_id) {
+	case IPC4_COPIER_MODULE_CFG_PARAM_LLP_READING:
+		if (!cd->endpoint || *data_offset < sizeof(struct ipc4_llp_reading))
+			return -EINVAL;
```
Comment (lgirdwood) [inline src/audio/copier.c:742]: lets have some trace err here

File: src/audio/copier.c [hunk 4]
```diff
@@ -719,6 +719,84 @@ static int copier_set_large_config(struct comp_dev *dev, uint32_t param_id,
...
+
+		if (comp_get_endpoint_type(cd->endpoint) != COMP_ENDPOINT_DAI)
+			return -EINVAL;
```
Comment (lgirdwood) [inline src/audio/copier.c:745]: and here.

File: src/audio/copier.c [hunk 5]
```diff
@@ -719,6 +719,84 @@ static int copier_set_large_config(struct comp_dev *dev, uint32_t param_id,
...
+
+		divide_u64_to_u32s(posn.comp_posn, &llp.llp_l, &llp.llp_u);
+		divide_u64_to_u32s(posn.wallclock, &llp.wclk_l, &llp.wclk_u);
+		memcpy_s(data, sizeof(llp), &llp, sizeof(llp));
+
+		return 0;
+
+	case IPC4_COPIER_MODULE_CFG_PARAM_LLP_READING_EXTENDED:
+		if (!cd->endpoint || *data_offset < sizeof(struct ipc4_llp_reading_extended))
+			return -EINVAL;
```
Comment (lgirdwood) [inline src/audio/copier.c:766]: and here

File: src/audio/copier.c [hunk 6]
```diff
@@ -719,6 +719,84 @@ static int copier_set_large_config(struct comp_dev *dev, uint32_t param_id,
...
+
+		if (comp_get_endpoint_type(cd->endpoint) != COMP_ENDPOINT_DAI)
+			return -EINVAL;
```
Comment (lgirdwood) [inline src/audio/copier.c:769]: and here (all the trace will help the driver folks).

File: src/ipc/ipc4/handler.c [hunk 7]
```diff
@@ -601,6 +601,8 @@ static int ipc4_process_module_message(union ipc4_message_header *ipc4)
 		break;
 	case SOF_IPC4_MOD_CONFIG_GET:
 	case SOF_IPC4_MOD_CONFIG_SET:
+		ret = IPC4_UNAVAILABLE;
+		break;
```
Comment (lgirdwood) [inline src/ipc/ipc4/handler.c:605]: trace_info would be good here.

**Review by lgirdwood — COMMENTED** [review]
overall looks good, just some minor stuff.

File: src/include/sof/lib/dma.h (hunk 1 above)
Comment (RanderWang) [reply src/include/sof/lib/dma.h:153]: sure

File: src/audio/copier.c (hunk 2 above)
Comment (RanderWang) [reply src/audio/copier.c:722]: sure

File: src/audio/copier.c (hunk 3 above)
Comment (RanderWang) [reply src/audio/copier.c:742]: sure

File: src/ipc/ipc4/handler.c (hunk 7 above)
Comment (RanderWang) [reply src/ipc/ipc4/handler.c:605]: sure

**Comment by keqiaozhang** [comment]
SOFCI TEST

**Comment by keqiaozhang** [comment]
SOFCI TEST

**Comment by keqiaozhang** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@RanderWang can you check CI
```
[ 91%] Building C object CMakeFiles/sof.dir/src/drivers/interrupt.c.o
/home/sof/work/sof.git/src/drivers/dw/dma.c: In function 'dw_dma_status':
/home/sof/work/sof.git/src/drivers/dw/dma.c:469:3: error: implicit declaration of function 'platform_dw_dma_llp_read'; did you mean 'platform_dw_dma_llp_enable'? [-Werror=implicit-function-declaration]
  469 |   platform_dw_dma_llp_read(channel->dma, channel, llp, llp + 1);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~
      |   platform_dw_dma_llp_enable
[ 91%] Building C object CMakeFiles/sof.dir/src/trace/dma-trace.c.o
```

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@RanderWang oh, maybe ignore my CI comment, I see you have just re pushed an update.

**Comment by RanderWang** [comment]
updated. I don't know why zephyr was failed

**Comment by lgirdwood** [comment]
> updated. I don't know why zephyr was failed

Known issue unrelated.

**Pull request merged**